bpo-38368: Added fix for ctypes crash when handling arrays in structs…

vsajip opened the pull request:
…/unions.

<!-- issue-number: [bpo-38368](https://bugs.python.org/issue38368) -->
https://bugs.python.org/issue38368
<!-- /issue-number -->

vsajip committed:
bpo-38368: Added fix for ctypes crash when handling arrays in structs/unions.

vsajip committed:
Simplified handling of ctypes structs/unions containing arrays.

ZackerySpytz commented on Modules/_ctypes/stgdict.c:
```diff
@@ -695,14 +721,17 @@ PyCStructUnionType_update_stgdict(PyObject *type, PyObject *fields, int isStruct
+            if (PyCArrayTypeObject_Check(desc)) {
                 Py_ssize_t length = dict->length;
                 StgDictObject *edict;
+                ffi_type * dummy_struct;
```
```suggestion
                ffi_type *dummy_struct;
```

ZackerySpytz commented on Modules/_ctypes/stgdict.c:
```diff
@@ -695,14 +721,17 @@ PyCStructUnionType_update_stgdict(PyObject *type, PyObject *fields, int isStruct
                 Py_ssize_t length = dict->length;
                 StgDictObject *edict;
+                ffi_type * dummy_struct;
+                ffi_type ** dummy_fields;
```
```suggestion
                ffi_type **dummy_fields;
```

ZackerySpytz commented on Modules/_ctypes/stgdict.c:
```diff
@@ -712,37 +741,36 @@ PyCStructUnionType_update_stgdict(PyObject *type, PyObject *fields, int isStruct
+                /* Do separate allocations for now. */
+                dummy_struct = PyMem_New(ffi_type, 1);
+                if (dummy_struct == NULL) {
+                    PyErr_NoMemory();
```
I think there should be a `Py_DECREF(pair);` here.

ZackerySpytz commented on Modules/_ctypes/stgdict.c:
```diff
@@ -712,37 +741,36 @@ PyCStructUnionType_update_stgdict(PyObject *type, PyObject *fields, int isStruct
+                }
+                dummy_fields = PyMem_New(ffi_type *, length + 1);
+                if (dummy_fields == NULL) {
+                    PyErr_NoMemory();
```
I think `dummy_struct` should be freed here. Also, `pair` should be decrefed.

vsajip replied:
Well, the problem is that we're doing these allocations in a loop, and unless we keep some sort of linked list or dynamically allocated structure to keep track of all our allocations (both `dummy_struct` and `dummy_fields` in earlier iterations of the loop), we will leak this memory in an Out Of Memory (OOM) situation. I'm not aware that we currently do this type of recovery for these internal allocations in an OOM situation, but happy to be proved wrong.

vsajip replied:
I will have a think about an alternative allocation strategy which allows an easier recovery, but it will be potentially a little harder to follow the logic compared to what's there now.

vsajip committed:
Added some more test logic.

vsajip committed:
Changes in response to review comments.

vsajip committed:
Reverted to single-allocation strategy for easier backing out on failure.

vsajip committed:
Amended incorrect assertion.

vsajip committed:
Moved assertion out of condition.

zooba reviewed (approved):
I'd like to have one more person go over this line-by-line as well to track the memory usage, but I didn't spot anything that seemed wrong.

ericsnowcurrently commented on Modules/_ctypes/stgdict.c:
```diff
@@ -681,24 +719,116 @@ PyCStructUnionType_update_stgdict(PyObject *type, PyObject *fields, int isStruct
+            if (!PyArg_ParseTuple(pair, "UO|i", &name, &desc, &bitsize)) {
+                PyErr_SetString(PyExc_TypeError,
+                    "'_fields_' must be a sequence of (name, C type) pairs");
+                Py_XDECREF(pair);
```
Why not use `Py_DECREF()`?  Don't we already know that `pair` isn't NULL?

vsajip replied:
Good point, I'll change that.

ericsnowcurrently commented on Modules/_ctypes/stgdict.c:
```diff
@@ -681,24 +719,116 @@ PyCStructUnionType_update_stgdict(PyObject *type, PyObject *fields, int isStruct
+                PyMem_Free(type_block);
+                return -1;
+            }
             if (!PyArg_ParseTuple(pair, "UO|i", &name, &desc, &bitsize)) {
```
This check (but certainly not the call) might be redundant based on the earlier for loop.

vsajip replied:
Although it _shouldn't_ fail, I suppose one can't guarantee it? For that reason, I'd leave the check in, it makes the code a wee bit bigger but shouldn't really impact performance.

ericsnowcurrently commented on Modules/_ctypes/stgdict.c:
```diff
@@ -681,24 +719,116 @@ PyCStructUnionType_update_stgdict(PyObject *type, PyObject *fields, int isStruct
                 return -1;
             }
             dict = PyType_stgdict(desc);
             if (dict == NULL) {
```
This check seems redundant based on the earlier for loop.

vsajip replied:
As per earlier comment - would prefer to leave the check in for now.

ericsnowcurrently reviewed (approved):
LGTM.  I've left some minor comments, but they aren't critical.

I did double-check the memory allocations and everything looks good.  The only allocations were the 2 `PySequence_GetItem()` calls and the one `PyMem_Malloc()` call.  They are cleaned up in all the right spots.

vsajip committed:
Changed the name of a #defined constant to be more descriptive,
changed a couple of Py_XDECREF to Py_DECREF, and added some comments
about possibly defensive checks.
